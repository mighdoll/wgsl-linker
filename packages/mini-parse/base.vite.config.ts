import { UserConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import dts from "vite-plugin-dts";
import { resolve } from "path";

export function baseViteConfig(): UserConfig {
  return {
    plugins: [
      tsconfigPaths(),
      dts(), // generate .d.ts files
    ],
    build: {
      lib: {
        name: "mini-parse",
        entry: [resolve(__dirname, "src/index.ts")],
      },
      minify: false,
      sourcemap: true,
    },
  };
}
