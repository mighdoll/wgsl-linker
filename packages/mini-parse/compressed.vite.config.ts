/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
// import { visualizer } from "rollup-plugin-visualizer";
import { baseViteConfig } from "./vite-base-config.js";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
