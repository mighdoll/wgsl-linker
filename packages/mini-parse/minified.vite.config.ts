/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
// import { visualizer } from "rollup-plugin-visualizer";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
