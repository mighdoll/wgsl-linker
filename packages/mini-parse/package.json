{
  "name": "mini-parse",
  "version": "0.2.0",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "pr": "printenv",
    "build": "run-s build:main build:util build:compressed",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:compressed": "vite -c compressed.vite.config.ts build",
    "build:brotli": "brotli dist/minified.cjs",
    "test": "vitest",
    "organize": "organize-imports-cli tsconfig.json",
    "format": "prettier-eslint \"src/**/*.ts\" --write",
    "lint": "eslint src"
  },
  "type": "module",
  "repository": "github:mighdoll/wgsl-linker",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts"
    },
    "./minified": {
      "import": "./dist/minified.js",
      "require": "./dist/minified.cjs",
      "types": "./dist/index.d.ts"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^20.11.16",
    "@typescript-eslint/eslint-plugin": "^6.21.0",
    "@typescript-eslint/parser": "^6.21.0",
    "@vitest/ui": "^1.1.1",
    "berry-pretty": "^0.0.4",
    "brotli": "^1.3.3",
    "esbuild": "^0.20.0",
    "eslint": "^8.56.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "terser": "^5.27.0",
    "typescript": "^5.3.3",
    "vite": "^5.0.12",
    "vite-plugin-dts": "^3.6.4",
    "vite-tsconfig-paths": "^4.2.1",
    "vitest": "^1.2.2"
  }
}
