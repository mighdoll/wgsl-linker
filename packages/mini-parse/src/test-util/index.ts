export * from "./LogCatcher.js";
export * from "./TestParse.js";
