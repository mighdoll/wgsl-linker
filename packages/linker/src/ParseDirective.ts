import {
  Parser,
  anyThrough,
  kind,
  opt,
  or,
  repeat,
  req,
  seq,
  setTraceName,
  tokens,
  tracing,
  withSep,
} from "mini-parse";
import {
  ExportElem,
  ImportElem,
  ExtendsElem,
  NamedElem,
} from "./AbstractElems.js";
import {
  argsTokens,
  lineCommentTokens,
  mainTokens,
  moduleTokens,
} from "./MatchWgslD.js";
import { eolf, makeElem } from "./ParseSupport.js";

/* parse #directive enhancements to wgsl: #import, #export, etc. */

const argsWord = kind(argsTokens.arg);

// prettier-ignore
/** ( <a> <,b>* )  with optional comments interspersed, does not span lines */
export const directiveArgs: Parser<string[]> = 
  seq(
    "(", 
    withSep(",", argsWord), 
    req(")")
  ).map((r) => r.value[1]);

/** foo <(A,B)> <as boo> <from bar>  EOL */
function importPhrase<T extends ImportElem | ExtendsElem>(
  kind: T["kind"]
): Parser<T> {
  const p = seq(
    argsWord.named("name"),
    opt(directiveArgs.named("args")),
    opt(seq("as", argsWord.named("as"))),
    opt(seq("from", argsWord.named("from")))
  ).map((r) => {
    // flatten 'args' by putting it with the other extracted names
    const named: (keyof T)[] = ["name", "from", "as", "args"];
    return makeElem<T>(kind, r, named, []);
  });

  return p;
}

const importElemPhrase = importPhrase<ImportElem>("import");
const importMergeElemPhrase = importPhrase<ExtendsElem>("extends");

export const importing = seq(
  "importing",
  seq(importElemPhrase.named("importing")),
  repeat(seq(",", importElemPhrase.named("importing")))
);

/** #import foo <(a,b)> <as boo> <from bar>  EOL */
const importDirective = seq(
  "#import",
  seq(importElemPhrase.named("i"), eolf)
).map((r) => {
  const imp: ImportElem = r.named.i[0];
  imp.start = r.start; // use start of #import, not import phrase
  r.app.state.push(imp);
});

const importMergeSym = Symbol("extends");

export const importMergeDirective = seq(
  "#extends",
  seq(importMergeElemPhrase.named(importMergeSym), eolf)
).map((r) => {
  const imp: ExtendsElem = r.named[importMergeSym][0];
  imp.start = r.start; // use start of #import, not import phrase
  r.app.state.push(imp);
});

/** #export <foo> <(a,b)> <importing bar(a) <zap(b)>* > EOL */
export const exportDirective = seq(
  "#export",
  seq(opt(directiveArgs.named("args")), opt(importing), eolf)
).map((r) => {
  // flatten 'args' by putting it with the other extracted names
  const e = makeElem<ExportElem>("export", r, ["args"], ["importing"]);
  r.app.state.push(e);
});

const moduleDirective = oneArgDirective("module");

const templateDirective = oneArgDirective("template");

function oneArgDirective<T extends NamedElem>(
  elemKind: T["kind"]
): Parser<void> {
  return seq(
    `#${elemKind}`,
    tokens(moduleTokens, req(kind(moduleTokens.moduleName).named("name"))),
    eolf
  ).map((r) => {
    const e = makeElem<T>(elemKind, r, ["name"]);
    r.app.state.push(e);
  });
}

export const directive = tokens(
  argsTokens,
  seq(
    repeat("\n"),
    or(
      exportDirective,
      importDirective,
      importMergeDirective,
      moduleDirective,
      templateDirective
    )
  )
);

const skipToEol = tokens(lineCommentTokens, anyThrough(eolf));

/** parse a line comment possibly containg a #directive
 *    // <#import|#export|any>
 * if a directive is found it is handled internally (e.g.
 * by pushing an AbstractElem to the app context) */
export const lineCommentOptDirective = seq(
  tokens(mainTokens, "//"),
  or(directive, skipToEol)
);

// enableTracing();
if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    directiveArgs,
    importElemPhrase,
    importMergeElemPhrase,
    importing,
    importDirective,
    importMergeDirective,
    exportDirective,
    lineCommentOptDirective,
    moduleDirective,
    templateDirective,
    directive,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
