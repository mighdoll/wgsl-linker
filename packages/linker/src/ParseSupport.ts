import {
  ExtendedResult,
  Parser,
  any,
  anyNot,
  disablePreParse,
  fn,
  kind,
  makeEolf,
  or,
  repeat,
  req,
  resultLog,
  seq,
  setTraceName,
  tracing,
  withSep,
} from "mini-parse";
import { AbstractElem } from "./AbstractElems.js";
import { argsTokens, mainTokens } from "./MatchWgslD.js";
import { lineCommentOptDirective } from "./ParseDirective.js";

/* Basic parsing functions for comment handling, eol, etc. */

export const word = kind(mainTokens.word);
export const wordNum = or(word, kind(mainTokens.digits));

export const unknown = any().map((r) => {
  const { kind, text } = r.value;
  resultLog(r, `??? ${kind}: '${text}'`);
});

export const blockComment: Parser<any> = seq(
  "/*",
  repeat(or(() => blockComment, anyNot("*/"))),
  req("*/")
);

export const comment = or(
  fn(() => lineCommentOptDirective),
  blockComment
);

export const eolf: Parser<any> = disablePreParse(
  makeEolf(argsTokens, argsTokens.ws)
);

/** ( a1, b1* ) with optinoal comments, spans lines */
export const wordNumArgs: Parser<string[]> = seq(
  "(",
  withSep(",", wordNum),
  req(")")
).map((r) => r.value[1]);

/** create an AbstractElem from parse results
 * @param named keys in the named result to copy to
 *  like named fields in the abstract elem (as a single value)
 * @param namedArray keys in the named result to copy to
 *  like named fields in the abstract elem (as an array)
 */
export function makeElem<U extends AbstractElem>(
  kind: U["kind"],
  er: ExtendedResult<any>,
  named: (keyof U)[] = [],
  namedArrays: (keyof U)[] = []
): U {
  const { start, end } = er;
  const nv = mapIfDefined(named, er.named as NameRecord<U>, true);
  const av = mapIfDefined(namedArrays, er.named as NameRecord<U>);
  return { kind, start, end, ...nv, ...av } as U;
}

type NameRecord<A> = Record<keyof A, string[]>;

function mapIfDefined<A>(
  keys: (keyof A)[],
  array: Record<keyof A, string[]>,
  firstElem?: boolean
): Partial<Record<keyof A, string>> {
  const entries = keys.flatMap((k) => {
    const ak = array[k];
    const v = firstElem ? ak?.[0] : ak;

    if (v === undefined) return [];
    else return [[k, v]];
  });
  return Object.fromEntries(entries);
}

// enableTracing();
if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    skipBlockComment: blockComment,
    comment,
    wordNumArgs,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
