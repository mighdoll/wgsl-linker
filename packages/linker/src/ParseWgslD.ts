import {
  anyNot,
  anyThrough,
  eof,
  fn,
  kind,
  matchingLexer,
  opt,
  or,
  Parser,
  ParserContext,
  ParserInit,
  preParse,
  repeat,
  req,
  seq,
  setTraceName,
  simpleParser,
  SrcMap,
  tracing,
  withSep,
} from "mini-parse";
import {
  AbstractElem,
  AliasElem,
  CallElem,
  FnElem,
  FnNameElem,
  GlobalDirectiveElem,
  StructElem,
  StructMemberElem,
  TypeNameElem,
  TypeRefElem,
  VarElem,
} from "./AbstractElems.js";
import { mainTokens } from "./MatchWgslD.js";
import { directive } from "./ParseDirective.js";
import {
  comment,
  makeElem,
  unknown,
  word,
  wordNumArgs,
} from "./ParseSupport.js";

/** parser that recognizes key parts of WGSL and also directives like #import */

// prettier gets confused if we leave the quoted parens inline so make consts for them here
const lParen = "(";
const rParen = ")";

export interface ParseState {
  ifStack: boolean[]; // stack used while processiing nested #if #else #endif directives
  params: Record<string, any>; // user provided params to templates, code gen and #if directives
}

const optAttributes = repeat(seq(kind(mainTokens.attr), opt(wordNumArgs)));
const possibleTypeRef = Symbol("typeRef");

const globalDirectiveOrAssert = seq(
  or("diagnostic", "enable", "requires", "const_assert"),
  req(anyThrough(";"))
).map((r) => {
  const e = makeElem<GlobalDirectiveElem>("globalDirective", r);
  r.app.state.push(e);
});

/** parse an identifier into a TypeNameElem */
export const typeNameDecl = req(word.named("name")).map((r) => {
  return makeElem<TypeNameElem>("typeName", r, ["name"]);
});

/** parse an identifier into a TypeNameElem */
export const fnNameDecl = req(word.named("name"), "missing fn name").map(
  (r) => {
    return makeElem<FnNameElem>("fnName", r, ["name"]);
  }
);

/** find possible references to user types (structs) in this possibly nested template */
export const template: Parser<any> = seq(
  "<",
  or(
    word.named(possibleTypeRef), // only the first element of the template can be a type
    fn(() => template)
  ),
  repeat(
    or(
      fn(() => template),
      anyNot(">") // we don't care about the rest of the template
    )
  ),
  req(">")
);

/** find possible references to user structs in this type specifier and any templates */
export const typeSpecifier: Parser<TypeRefElem[]> = seq(
  word.named(possibleTypeRef),
  opt(template)
).map((r) =>
  r.named[possibleTypeRef].map((name) => {
    const e = makeElem<TypeRefElem>("typeRef", r);
    e.name = name;
    return e;
  })
);

export const structMember = seq(
  optAttributes,
  word.named("name"),
  ":",
  req(typeSpecifier.named("typeRefs"))
).map((r) => {
  return makeElem<StructMemberElem>("member", r, ["name", "typeRefs"]);
});

export const structDecl = seq(
  "struct",
  req(typeNameDecl).named("nameElem"),
  req("{"),
  withSep(",", structMember).named("members"),
  req("}")
).map((r) => {
  const e = makeElem<StructElem>("struct", r, ["members"]);
  const nameElem = r.named.nameElem[0];
  e.nameElem = nameElem;
  e.name = nameElem.name;
  r.app.state.push(e);
});

// keywords that can be followed by (), not to be confused with fn calls
const callishKeyword = simpleParser("keyword", (ctx: ParserContext) => {
  const keywords = ["if", "for", "while", "const_assert", "return"];
  const token = ctx.lexer.next();
  const text = token?.text;
  if (text && keywords.includes(text)) {
    return text;
  }
});

export const fnCall = seq(
  word
    .named("name")
    .map((r) => makeElem<CallElem>("call", r, ["name"]))
    .named("calls"), // we collect this in fnDecl, to attach to FnElem
  "("
);

// prettier-ignore
const fnParam = seq(
  optAttributes,
  word,
  opt(seq(":", req(typeSpecifier.named("typeRefs"))))
);

const fnParamList = seq(lParen, withSep(",", fnParam), rParen);

// prettier-ignore
const variableDecl = seq(
  or("const", "var", "let", "override"), 
  word, 
  ":", 
  req(typeSpecifier).named("typeRefs")
);

const block: Parser<any> = seq(
  "{",
  repeat(
    or(
      callishKeyword,
      fnCall,
      fn(() => block),
      variableDecl,
      anyNot("}")
    )
  ),
  req("}")
);

export const fnDecl = seq(
  optAttributes,
  "fn",
  req(fnNameDecl).named("nameElem"),
  // TODO some built in functions can have a template here, e.g. bitcast
  req(fnParamList),
  opt(seq("->", optAttributes, typeSpecifier.named("typeRefs"))),
  req(block)
).map((r) => {
  const e = makeElem<FnElem>("fn", r);
  const nameElem = r.named.nameElem[0];
  e.nameElem = nameElem;
  e.name = nameElem.name;
  e.calls = r.named.calls || [];
  e.typeRefs = r.named.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalVar = seq(
  optAttributes,
  or("const", "override", "var"),
  opt(template),
  word.named("name"),
  opt(seq(":", req(typeSpecifier.named("typeRefs")))),
  req(anyThrough(";"))
).map((r) => {
  const e = makeElem<VarElem>("var", r, ["name"]);
  e.typeRefs = r.named.typeRefs?.flat() || [];
  r.app.state.push(e);
});

export const globalAlias = seq(
  "alias",
  req(word.named("name")),
  req("="),
  req(typeSpecifier).named("typeRefs"),
  req(";")
).map((r) => {
  const e = makeElem<AliasElem>("alias", r, ["name", "typeRefs"]);
  r.app.state.push(e);
});

const globalDecl = or(fnDecl, globalVar, globalAlias, structDecl, ";");

const rootDecl = or(globalDirectiveOrAssert, globalDecl, directive, unknown);

const root = preParse(comment, seq(repeat(rootDecl), eof()));

export function parseWgslD(
  src: string,
  srcMap?: SrcMap,
  params: Record<string, any> = {},
  maxParseCount: number | undefined = undefined,
  grammar = root
): AbstractElem[] {
  const lexer = matchingLexer(src, mainTokens);
  const state: AbstractElem[] = [];
  const context: ParseState = { ifStack: [], params };
  const app = {
    context,
    state,
  };
  const init: ParserInit = {
    lexer,
    app,
    srcMap,
    maxParseCount,
  };

  grammar.parse(init);

  return app.state;
}

if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    globalDirectiveOrAssert,
    template,
    typeSpecifier,
    structMember,
    structDecl,
    fnCall,
    fnParam,
    fnParamList,
    block,
    fnDecl,
    globalVar,
    globalAlias,
    globalDecl,
    rootDecl,
    root,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
