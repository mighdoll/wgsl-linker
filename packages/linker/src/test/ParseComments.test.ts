import { preParse } from "mini-parse";
import { expectNoLogErr } from "mini-parse/test-util";

import { expect, test } from "vitest";
import { lineCommentOptDirective } from "../ParseDirective.js";
import { blockComment, comment, wordNumArgs } from "../ParseSupport.js";
import { parseWgslD } from "../ParseWgslD.js";
import { testAppParse } from "./TestUtil.js";

test("lineComment parse // foo bar", () => {
  const src = "// foo bar";
  const { position } = testAppParse(lineCommentOptDirective, src);
  expect(position).eq(src.length);
});

test("lineComment parse // foo bar \\n", () => {
  const src = "// foo bar\n";
  const { position } = testAppParse(lineCommentOptDirective, src);
  expect(position).eq(src.length);
});

test("blockComment parses /* comment */", () => {
  const src = "/* comment */";
  expectNoLogErr(() => {
    const { parsed } = testAppParse(blockComment, src);
    expect(parsed).toMatchSnapshot();
  });
});

test("skipBlockComment parses nested comment", () => {
  const src = "/** comment1 /* comment2 */ */";
  expectNoLogErr(() => {
    testAppParse(blockComment, src);
  });
});

test("parse fn with line comment", () => {
  const src = `
    fn binaryOp() { // binOpImpl
    }`;
  const parsed = parseWgslD(src);
  expect(parsed).toMatchSnapshot();
});

test("wordNumArgs parses (a, b, 1) with line comments everywhere", () => {
  const src = `(
    // aah
    a, 
    // boh
    b, 
    // oneness
    1
    // satsified
    )`;
  const { parsed } = testAppParse(preParse(comment, wordNumArgs), src);
  expect(parsed?.value).toMatchSnapshot();
});

test("parse empty line comment", () => {
  const src = `
  var workgroupThreads= 4;                          // 
  `;
  expectNoLogErr(() => parseWgslD(src));
});

test("parse line comment with #replace", () => {
  const src = ` 
  const workgroupThreads= 4;                          // #replace 4=workgroupThreads
  `;
  expectNoLogErr(() => parseWgslD(src));
});
