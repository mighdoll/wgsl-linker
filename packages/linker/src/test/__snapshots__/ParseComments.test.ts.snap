// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`blockComment parses /* comment */ 1`] = `
{
  "named": {},
  "value": [
    "/*",
    [
      {
        "kind": "word",
        "text": "comment",
      },
    ],
    "*/",
  ],
}
`;

exports[`parse fn with line comment 1`] = `
[
  {
    "calls": [],
    "end": 39,
    "kind": "fn",
    "name": "binaryOp",
    "nameElem": {
      "end": 16,
      "kind": "fnName",
      "name": "binaryOp",
      "start": 8,
    },
    "start": 5,
    "typeRefs": [],
  },
]
`;

exports[`wordNumArgs parses (a, b, 1) with line comments everywhere 1`] = `
[
  "a",
  "b",
  "1",
]
`;
